File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/PrivateKeys.kt
package com.blockchaincommons.bccomponents

import com.blockchaincommons.bctags.TAG_PRIVATE_KEYS
import com.blockchaincommons.bcur.UR
import com.blockchaincommons.bcur.URCodable
import com.blockchaincommons.dcbor.Cbor
import com.blockchaincommons.dcbor.CborCase
import com.blockchaincommons.dcbor.CborTaggedCodable
import com.blockchaincommons.dcbor.CborTaggedUtils
import com.blockchaincommons.dcbor.Tag
import com.blockchaincommons.dcbor.tagsForValues

/**
 * An interface for types that can provide a complete set of private cryptographic keys.
 *
 * Types implementing this interface can be used as a source of [PrivateKeys],
 * which contain both signing and encryption private keys.
 */
interface PrivateKeysProvider {
    /**
     * Returns a complete set of private keys for cryptographic operations.
     */
    fun privateKeys(): PrivateKeys
}

/**
 * A container combining signing and encapsulation private keys.
 *
 * [PrivateKeys] packages a [SigningPrivateKey] for creating digital signatures
 * with an [EncapsulationPrivateKey] for decrypting messages, providing a
 * complete private key set for cryptographic operations.
 *
 * This type is typically used alongside its public counterpart, [PublicKeys],
 * to enable secure communication between entities.
 */
class PrivateKeys(
    val signingPrivateKey: SigningPrivateKey,
    val encapsulationPrivateKey: EncapsulationPrivateKey,
) : PrivateKeysProvider, Signer, Decrypter, ReferenceProvider, CborTaggedCodable, URCodable {

    /**
     * Derives the corresponding [PublicKeys] from this private key set.
     */
    fun publicKeys(): PublicKeys =
        PublicKeys(
            signingPrivateKey.publicKey(),
            encapsulationPrivateKey.publicKey(),
        )

    // -- PrivateKeysProvider --

    override fun privateKeys(): PrivateKeys = this

    // -- Signer --

    override fun signWithOptions(message: ByteArray, options: SigningOptions?): Signature =
        signingPrivateKey.signWithOptions(message, options)

    // -- Decrypter --

    override fun encapsulationPrivateKey(): EncapsulationPrivateKey =
        encapsulationPrivateKey

    override fun decapsulateSharedSecret(ciphertext: EncapsulationCiphertext): SymmetricKey =
        encapsulationPrivateKey.decapsulateSharedSecret(ciphertext)

    // -- ReferenceProvider --

    override fun reference(): Reference =
        Reference.fromDigest(Digest.fromImage(taggedCbor().toCborData()))

    // -- equals / hashCode --

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PrivateKeys) return false
        return signingPrivateKey == other.signingPrivateKey &&
            encapsulationPrivateKey == other.encapsulationPrivateKey
    }

    override fun hashCode(): Int {
        var result = signingPrivateKey.hashCode()
        result = 31 * result + encapsulationPrivateKey.hashCode()
        return result
    }

    override fun toString(): String =
        "PrivateKeys(${refHexShort()}, $signingPrivateKey, $encapsulationPrivateKey)"

    // -- CBOR --

    override fun cborTags(): List<Tag> = tagsForValues(listOf(TAG_PRIVATE_KEYS))

    /**
     * Encodes as a CBOR array: [signing_key_tagged_cbor, encapsulation_key_tagged_cbor].
     */
    override fun untaggedCbor(): Cbor {
        val signingKeyCbor = signingPrivateKey.taggedCbor()
        val encapsulationKeyCbor = encapsulationPrivateKey.taggedCbor()
        return Cbor.fromArray(listOf(signingKeyCbor, encapsulationKeyCbor))
    }

    companion object {
        /** Decodes [PrivateKeys] from untagged CBOR (an array). */
        fun fromUntaggedCbor(cbor: Cbor): PrivateKeys {
            val case = cbor.cborCase
            if (case !is CborCase.Array) {
                throw BcComponentsException.invalidData("PrivateKeys", "expected array")
            }
            val elements = case.value
            if (elements.size != 2) {
                throw BcComponentsException.invalidData(
                    "PrivateKeys", "expected array of length 2, got ${elements.size}",
                )
            }
            val signingPrivateKey = SigningPrivateKey.fromTaggedCbor(elements[0])
            val encapsulationPrivateKey = EncapsulationPrivateKey.fromTaggedCbor(elements[1])
            return PrivateKeys(signingPrivateKey, encapsulationPrivateKey)
        }

        /** Decodes [PrivateKeys] from tagged CBOR. */
        fun fromTaggedCbor(cbor: Cbor): PrivateKeys =
            CborTaggedUtils.fromTaggedCbor(
                cbor,
                tagsForValues(listOf(TAG_PRIVATE_KEYS)),
            ) { fromUntaggedCbor(it) }

        /** Decodes [PrivateKeys] from tagged CBOR binary data. */
        fun fromTaggedCborData(data: ByteArray): PrivateKeys =
            CborTaggedUtils.fromTaggedCborData(
                data,
                tagsForValues(listOf(TAG_PRIVATE_KEYS)),
            ) { fromUntaggedCbor(it) }

        /** Decodes [PrivateKeys] from a UR. */
        fun fromUr(ur: UR): PrivateKeys {
            ur.checkType("crypto-prvkeys")
            return fromUntaggedCbor(ur.cbor)
        }

        /** Decodes [PrivateKeys] from a UR string. */
        fun fromUrString(urString: String): PrivateKeys =
            fromUr(UR.fromUrString(urString))
    }
}

File: csharp/BCComponents/BCComponents/PrivateKeys.cs
using BlockchainCommons.BCTags;
using BlockchainCommons.DCbor;

namespace BlockchainCommons.BCComponents;

/// <summary>
/// A container combining signing and encapsulation private keys.
/// </summary>
/// <remarks>
/// <para>
/// <see cref="PrivateKeys"/> packages a <see cref="SigningPrivateKey"/> for creating
/// digital signatures with an <see cref="EncapsulationPrivateKey"/> for decrypting
/// messages, providing a complete private key set for cryptographic operations.
/// </para>
/// <para>
/// This type is typically used alongside its public counterpart, <see cref="PublicKeys"/>,
/// to enable secure communication between entities.
/// </para>
/// </remarks>
public sealed class PrivateKeys
    : IEquatable<PrivateKeys>,
      IPrivateKeysProvider,
      ISigner,
      IDecrypter,
      IReferenceProvider,
      ICborTaggedEncodable,
      ICborTaggedDecodable
{
    /// <summary>Gets the signing private key.</summary>
    public SigningPrivateKey SigningPrivateKey { get; }

    /// <summary>Gets the encapsulation private key.</summary>
    public EncapsulationPrivateKey EncapsulationPrivateKey { get; }

    /// <summary>
    /// Creates a new <see cref="PrivateKeys"/> from the given signing and encapsulation private keys.
    /// </summary>
    /// <param name="signingPrivateKey">The signing private key.</param>
    /// <param name="encapsulationPrivateKey">The encapsulation private key.</param>
    public PrivateKeys(SigningPrivateKey signingPrivateKey, EncapsulationPrivateKey encapsulationPrivateKey)
    {
        SigningPrivateKey = signingPrivateKey;
        EncapsulationPrivateKey = encapsulationPrivateKey;
    }

    /// <summary>
    /// Derives the corresponding <see cref="PublicKeys"/> from this private key set.
    /// </summary>
    /// <returns>The corresponding <see cref="PublicKeys"/>.</returns>
    public PublicKeys PublicKeys() =>
        new(
            SigningPrivateKey.PublicKey(),
            EncapsulationPrivateKey.PublicKey());

    // --- IPrivateKeysProvider ---

    /// <inheritdoc/>
    PrivateKeys IPrivateKeysProvider.PrivateKeys() => this;

    // --- ISigner ---

    /// <inheritdoc/>
    public Signature SignWithOptions(byte[] message, SigningOptions? options = null) =>
        SigningPrivateKey.SignWithOptions(message, options);

    // --- IDecrypter ---

    /// <inheritdoc/>
    EncapsulationPrivateKey IDecrypter.EncapsulationPrivateKey() => EncapsulationPrivateKey;

    /// <inheritdoc/>
    SymmetricKey IDecrypter.DecapsulateSharedSecret(EncapsulationCiphertext ciphertext) =>
        EncapsulationPrivateKey.DecapsulateSharedSecret(ciphertext);

    // --- IReferenceProvider ---

    /// <inheritdoc/>
    public Reference Reference() =>
        BCComponents.Reference.FromDigest(Digest.FromImage(TaggedCbor().ToCborData()));

    // --- IEquatable<PrivateKeys> ---

    /// <inheritdoc/>
    public bool Equals(PrivateKeys? other)
    {
        if (other is null) return false;
        return SigningPrivateKey == other.SigningPrivateKey
            && EncapsulationPrivateKey == other.EncapsulationPrivateKey;
    }

    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is PrivateKeys k && Equals(k);

    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(SigningPrivateKey, EncapsulationPrivateKey);

    /// <summary>Tests equality of two PrivateKeys instances.</summary>
    public static bool operator ==(PrivateKeys? left, PrivateKeys? right) =>
        left is null ? right is null : left.Equals(right);

    /// <summary>Tests inequality of two PrivateKeys instances.</summary>
    public static bool operator !=(PrivateKeys? left, PrivateKeys? right) => !(left == right);

    // --- ICborTagged ---

    /// <summary>Returns the CBOR tags for this type (40013).</summary>
    public static IReadOnlyList<Tag> CborTags =>
        GlobalTags.TagsForValues(BcTags.TagPrivateKeys);

    // --- ICborTaggedEncodable ---

    /// <summary>
    /// Returns the untagged CBOR representation as a two-element array
    /// [signing_key_tagged_cbor, encapsulation_key_tagged_cbor].
    /// </summary>
    public Cbor UntaggedCbor()
    {
        var elements = new List<Cbor>
        {
            SigningPrivateKey.TaggedCbor(),
            EncapsulationPrivateKey.ToCbor(),
        };
        return Cbor.FromList(elements);
    }

    /// <summary>Returns the tagged CBOR representation.</summary>
    public Cbor TaggedCbor() => Cbor.ToTaggedValue(CborTags[0], UntaggedCbor());

    /// <summary>Returns the tagged CBOR representation.</summary>
    public Cbor ToCbor() => TaggedCbor();

    // --- ICborTaggedDecodable ---

    /// <summary>
    /// Decodes a <see cref="PrivateKeys"/> from untagged CBOR (a two-element array).
    /// </summary>
    /// <param name="cbor">The untagged CBOR value.</param>
    /// <returns>A new <see cref="PrivateKeys"/>.</returns>
    public static PrivateKeys FromUntaggedCbor(Cbor cbor)
    {
        var elements = cbor.TryIntoArray();
        if (elements.Count != 2)
            throw BCComponentsException.InvalidData("PrivateKeys",
                $"expected array of length 2, got {elements.Count}");

        var signingPrivateKey = SigningPrivateKey.FromTaggedCbor(elements[0]);
        var encapsulationPrivateKey = BCComponents.EncapsulationPrivateKey.FromCbor(elements[1]);
        return new PrivateKeys(signingPrivateKey, encapsulationPrivateKey);
    }

    /// <summary>Decodes a <see cref="PrivateKeys"/> from tagged CBOR.</summary>
    /// <param name="cbor">The tagged CBOR value.</param>
    /// <returns>A new <see cref="PrivateKeys"/>.</returns>
    public static PrivateKeys FromTaggedCbor(Cbor cbor)
    {
        foreach (var tag in CborTags)
        {
            try
            {
                var item = cbor.TryIntoExpectedTaggedValue(tag);
                return FromUntaggedCbor(item);
            }
            catch (CborWrongTagException) { }
            catch (CborWrongTypeException) { }
        }
        throw new CborWrongTypeException();
    }

    // --- Display ---

    /// <inheritdoc/>
    public override string ToString() =>
        $"PrivateKeys({((IReferenceProvider)this).RefHexShort()}, {SigningPrivateKey}, {EncapsulationPrivateKey})";
}
